### PR TITLE
Numa_memdev_options: Assign proper host numa node automatically

### PR DESCRIPTION
**Pull request by wangxianxian**
1.Update function "check_host_numa_node_amount" to
  "get_host_numa_node" to check host numa node
  whose node size is non-zero
2.Assign host numa nodes whose node size is non-zero to
  "host-nodes" and "target_num"

ID:1716737
ID:1722351
Signed-off-by: Xianwang <xianwang@redhat.com>

### CODE REVIEW

**Comment by wangxianxian** [comment]
Test pass both on power8 and power9.
@MiriamDeng  @huangyum  If you are convenient, please help to review it, thanks

File: qemu/tests/numa_memdev_options.py [hunk 1]
```diff
@@ -4,21 +4,34 @@
...
+    #
+    #     :param test: QEMU test object
+    #     """
+    host_numa = utils_memory.numa_nodes()
+    node_list = []
+    if len(host_numa) < 2:
         test.cancel("The host numa nodes should be at least 2! But there is %d."
-                    % host_numa_nodes)
+                    % len(host_numa))
+    for i in range(len(host_numa)):
```
Comment (lolyu) [inline qemu/tests/numa_memdev_options.py]: One call of `numactl -H` is sufficient.

File: qemu/tests/numa_memdev_options.py [hunk 2]
```diff
@@ -4,21 +4,34 @@
...
+        node_size = int(process.getoutput(
+            "numactl -H | grep -w 'node %s' | grep size"
+            % str(host_numa[i])).split()[-2])
+        if node_size != 0:
+            node_list.append(host_numa[i])
+    if len(node_list) < 2:
+        test.cancel("The host numa nodes that whose size is not zero should be "
+                    "at least 2! But there is %d." % len(node_list))
+    print(node_list)
```
Comment (lolyu) [inline qemu/tests/numa_memdev_options.py]: redundant print

File: qemu/tests/numa_memdev_options.py [hunk 3]
```diff
@@ -4,21 +4,34 @@
...
+    host_numa = utils_memory.numa_nodes()
+    node_list = []
+    if len(host_numa) < 2:
         test.cancel("The host numa nodes should be at least 2! But there is %d."
-                    % host_numa_nodes)
+                    % len(host_numa))
+    for i in range(len(host_numa)):
+        node_size = int(process.getoutput(
+            "numactl -H | grep -w 'node %s' | grep size"
+            % str(host_numa[i])).split()[-2])
```
Comment (lolyu) [inline qemu/tests/numa_memdev_options.py]: no need to convert `host_numa[i]` to string.

File: qemu/tests/numa_memdev_mlock.py [hunk 4]
```diff
@@ -22,8 +22,15 @@ def run(test, params, env):
     :param params: Dictionary with the test parameters
     :param env: Dictionary with test environment.
     """
-    error_context.context("Check host's numa node(s) amount!", logging.info)
-    numa_memdev_options.check_host_numa_node_amount(test)
+    error_context.context("Check host's numa node(s)!", logging.info)
+    node_list_size_not_zero = numa_memdev_options.check_host_numa_node(test)
```
Comment (lolyu) [inline qemu/tests/numa_memdev_mlock.py]: why not put line 27 and 28 to the if block in line 30?

File: qemu/tests/numa_memdev_options.py (hunk 1 above)
Comment (wangxianxian) [reply qemu/tests/numa_memdev_options.py]: Thanks for lolyu's feedback, yes, once check is enough, I will update it.

File: qemu/tests/numa_memdev_options.py (hunk 2 above)
Comment (wangxianxian) [reply qemu/tests/numa_memdev_options.py]: Sorry, I forgot to remove it, just for debug

File: qemu/tests/numa_memdev_options.py (hunk 3 above)
Comment (wangxianxian) [reply qemu/tests/numa_memdev_options.py]: I agree it, if no need to sort size value, it is indeed not necessary to convert

File: qemu/tests/numa_memdev_mlock.py (hunk 4 above)
Comment (wangxianxian) [reply qemu/tests/numa_memdev_mlock.py]: I will update it.

**Comment by MiriamDeng** [comment]
I've jus have one concern here,initially,it's enough for us to set up huge page on one numa node but now it setups huge page on two numa nodes at this time,another one will be never used.Is it correct ? Thanks.
      - numa_ram_hugepage:
          backend_mem_mem0 = memory-backend-file
          backend_mem_mem1 = memory-backend-ram
          setup_hugepages = yes
          set_node_hugepage = yes
          mem-path_mem0 = /mnt/kvm_hugepage
          target_nodes = 0

**Comment by wangxianxian** [comment]
@MiriamDeng Thanks for your review, here, for "numa_ram_hugepage", setup hugepage on one numa node; for "numa_hugepage", setup hugepage on two numa nodes.

File: qemu/tests/numa_memdev_options.py [hunk 5]
```diff
@@ -1,24 +1,37 @@
...
-    host_numa_nodes = len(host_numa_nodes)
-    if host_numa_nodes < 2:
-        test.cancel("The host numa nodes should be at least 2! But there is %d."
-                    % host_numa_nodes)
+from virttest.test_setup import HugePageConfig
+
+
+def check_host_numa_node(test):
+    #     """
+    #     Check host NUMA node whose node size is not zero
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: I prefer to "Get host numa nodes ..."

File: qemu/tests/numa_memdev_options.py [hunk 6]
```diff
@@ -1,24 +1,37 @@
...
+    #
+    #     :param test: QEMU test object
+    #     """
+    host_numa = utils_memory.numa_nodes()
+    node_list = []
+    if len(host_numa) < 2:
+        test.cancel(
+            "The host numa nodes should be at least 2! But there is %d."
+            % len(host_numa))
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: I guess above if block can be dismissed since we have L31.

File: qemu/tests/numa_memdev_options.py [hunk 7]
```diff
@@ -111,10 +124,29 @@ def run(test, params, env):
     :param params: Dictionary with the test parameters
     :param env: Dictionary with test environment
     """
-    error_context.context("Check host's numa node(s) amount!", logging.info)
-    check_host_numa_node_amount(test)
-
+    error_context.context("Check host's numa node(s)!", logging.info)
+    node_list_size_not_zero = check_host_numa_node(test)
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: Maybe we can use "valid_nodes"?

File: qemu/tests/numa_memdev_options.py [hunk 8]
```diff
@@ -1,24 +1,37 @@
...
+    node_list = []
+    if len(host_numa) < 2:
+        test.cancel(
+            "The host numa nodes should be at least 2! But there is %d."
+            % len(host_numa))
+    numa_info = process.getoutput("numactl -H")
+    for i in host_numa:
+        node_size = re.findall(r"node %d size: \d+ \w" % i, numa_info)[0].split()[-2]
+        if node_size != '0':
+            node_list.append(str(i))
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: No need to use str here?

File: qemu/tests/numa_memdev_options.py (hunk 7 above)
Comment (huangyum) [reply qemu/tests/numa_memdev_options.py]: How about we move the if block in L31 out  of the function and put it here?

File: qemu/tests/numa_memdev_options.py [hunk 9]
```diff
@@ -111,10 +124,29 @@ def run(test, params, env):
...
+
+    if params.get('set_node_hugepage') == 'yes':
+        h = HugePageConfig(params)
+        hg_pz = str(h.get_hugepage_size())
+        h.set_node_num_huge_pages(num=params['target_num_node0'],
+                                  node=node1, pagesize=hg_pz)
+        if 'numa_hugepage' in params['shortname']:
+            h.set_node_num_huge_pages(num=params['target_num_node1'],
+                                      node=node2, pagesize=hg_pz)
+        h.mount_hugepage_fs()
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: Better to clean up the env after test.

File: qemu/tests/numa_memdev_options.py (hunk 5 above)
Comment (wangxianxian) [reply qemu/tests/numa_memdev_options.py]: Already update.

File: qemu/tests/numa_memdev_options.py (hunk 6 above)
Comment (wangxianxian) [reply qemu/tests/numa_memdev_options.py]: I have updated it

File: qemu/tests/numa_memdev_options.py (hunk 7 above)
Comment (wangxianxian) [reply qemu/tests/numa_memdev_options.py]: Yes, already update

**Comment by wangxianxian** [comment]
@huangyum @MiriamDeng @lolyu 
I have updated this patch as your comments, please review it again, thanks very much

File: qemu/tests/numa_memdev_options.py [hunk 10]
```diff
@@ -1,24 +1,30 @@
...
-    host_numa_nodes = utils_memory.numa_nodes()
-    host_numa_nodes = len(host_numa_nodes)
-    if host_numa_nodes < 2:
-        test.cancel("The host numa nodes should be at least 2! But there is %d."
-                    % host_numa_nodes)
+def get_host_numa_node(test):
+    #     """
+    #     Get host NUMA node whose node size is not zero
+    #
+    #     :param test: QEMU test object
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: Don't need the param "test".

Comment (wangxianxian) [reply qemu/tests/numa_memdev_options.py]: Ah, yes, I forgot to remove it here

File: qemu/tests/numa_memdev_options.py [hunk 11]
```diff
@@ -1,24 +1,28 @@
...
-    """
-    host_numa_nodes = utils_memory.numa_nodes()
-    host_numa_nodes = len(host_numa_nodes)
-    if host_numa_nodes < 2:
-        test.cancel("The host numa nodes should be at least 2! But there is %d."
-                    % host_numa_nodes)
+def get_host_numa_node():
+    #     """
+    #     Get host NUMA node whose node size is not zero
+    #     """
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: Don't need "#" in above lines.

File: qemu/tests/numa_memdev_options.py (hunk 9 above)
Comment (huangyum) [reply qemu/tests/numa_memdev_options.py]: If keep param "setup_hugepages = yes", hugepages could be set properly in preprocess_vm, then there is no need to set it here. Have tried on my host.

**Comment by MiriamDeng** [comment]
The job could pass
 (143/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_default.prealloc: PASS (67.50 s)
 (144/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_default.no_prealloc: PASS (29.63 s)
 (145/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_bind.prealloc: PASS (65.85 s)
 (146/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_bind.no_prealloc: PASS (71.24 s)
 (147/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_interleave.prealloc: PASS (70.00 s)
 (148/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_interleave.no_prealloc: PASS (63.34 s)
 (149/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_preferred.prealloc: PASS (64.37 s)
 (150/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_preferred.no_prealloc: PASS (63.20 s)
 (151/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_default.prealloc: PASS (64.33 s)
 (152/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_default.no_prealloc: PASS (64.25 s)
 (153/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_bind.prealloc: PASS (64.17 s)
 (154/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_bind.no_prealloc: PASS (64.10 s)
 (155/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_interleave.prealloc: PASS (63.09 s)
 (156/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_interleave.no_prealloc: PASS (69.95 s)
 (157/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_preferred.prealloc: PASS (65.88 s)
 (158/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_preferred.no_prealloc: PASS (64.04 s)
 (159/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_default.prealloc: PASS (64.35 s)
 (160/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_default.no_prealloc: PASS (63.55 s)
 (161/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_bind.prealloc: PASS (64.38 s)
 (162/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_bind.no_prealloc: PASS (65.85 s)
 (163/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_interleave.prealloc: PASS (63.52 s)
 (164/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_interleave.no_prealloc: PASS (64.26 s)
 (165/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_preferred.prealloc: PASS (63.44 s)
 (166/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_preferred.no_prealloc: PASS (63.29 s)
 (167/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_default.prealloc: PASS (66.46 s)
 (168/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_default.no_prealloc: PASS (64.47 s)
 (169/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_bind.prealloc: PASS (68.88 s)
 (170/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_bind.no_prealloc: PASS (63.33 s)
 (171/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_interleave.prealloc: PASS (63.41 s)
 (172/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_interleave.no_prealloc: PASS (68.79 s)
 (173/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_preferred.prealloc: PASS (71.30 s)
 (174/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_preferred.no_prealloc: PASS (65.51 s)

**Comment by MiriamDeng** [comment]
The jobs could pass
 (143/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_default.prealloc: PASS (67.50 s)
 (144/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_default.no_prealloc: PASS (29.63 s)
 (145/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_bind.prealloc: PASS (65.85 s)
 (146/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_bind.no_prealloc: PASS (71.24 s)
 (147/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_interleave.prealloc: PASS (70.00 s)
 (148/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_interleave.no_prealloc: PASS (63.34 s)
 (149/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_preferred.prealloc: PASS (64.37 s)
 (150/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_policy.policy_preferred.no_prealloc: PASS (63.20 s)
 (151/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_default.prealloc: PASS (64.33 s)
 (152/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_default.no_prealloc: PASS (64.25 s)
 (153/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_bind.prealloc: PASS (64.17 s)
 (154/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_bind.no_prealloc: PASS (64.10 s)
 (155/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_interleave.prealloc: PASS (63.09 s)
 (156/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_interleave.no_prealloc: PASS (69.95 s)
 (157/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_preferred.prealloc: PASS (65.88 s)
 (158/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_preferred.no_prealloc: PASS (64.04 s)
 (159/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_default.prealloc: PASS (64.35 s)
 (160/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_default.no_prealloc: PASS (63.55 s)
 (161/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_bind.prealloc: PASS (64.38 s)
 (162/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_bind.no_prealloc: PASS (65.85 s)
 (163/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_interleave.prealloc: PASS (63.52 s)
 (164/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_interleave.no_prealloc: PASS (64.26 s)
 (165/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_preferred.prealloc: PASS (63.44 s)
 (166/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_preferred.no_prealloc: PASS (63.29 s)
 (167/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_default.prealloc: PASS (66.46 s)
 (168/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_default.no_prealloc: PASS (64.47 s)
 (169/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_bind.prealloc: PASS (68.88 s)
 (170/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_bind.no_prealloc: PASS (63.33 s)
 (171/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_interleave.prealloc: PASS (63.41 s)
 (172/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_interleave.no_prealloc: PASS (68.79 s)
 (173/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_preferred.prealloc: PASS (71.30 s)
 (174/247) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.1.0.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_preferred.no_prealloc: PASS (65.51 s)

File: qemu/tests/numa_memdev_options.py [hunk 12]
```diff
@@ -111,10 +114,31 @@ def run(test, params, env):
...
+
+    if params.get('policy_mem') != 'default':
+        error_context.context("Assign host's numa node(s)!", logging.info)
+        params['host-nodes_mem0'] = node1
+        params['host-nodes_mem1'] = node2
+
+    if params.get('set_node_hugepage') == 'yes':
+        if 'numa_hugepage' in params['shortname']:
+            params['target_nodes'] = "%s %s" % (node1, node2)
+            params['target_num_node%s' % node2] = '1546'
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: Let's put the number in cfg, or just calculate the number by size_memX / Hugepagesize.
size_mem0 = 1024M
size_mem1 = 3072M

File: qemu/tests/numa_memdev_options.py [hunk 13]
```diff
@@ -111,10 +114,31 @@ def run(test, params, env):
...
+        else:
+            params['target_nodes'] = node1
+        params['target_num_node%s' % node1] = '522'
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: Same as above.

File: qemu/tests/numa_memdev_options.py [hunk 14]
```diff
@@ -1,24 +1,29 @@
...
 
 from avocado.utils import process
 
 from virttest import error_context
 from virttest import utils_misc
+from virttest import env_process
 from virttest.staging import utils_memory
 from virttest.compat_52lts import decode_to_text
+from virttest.utils_numeric import normalize_data_size
+from virttest.staging.utils_memory import get_huge_page_size
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: utils_memory has been imported in L9, why need import get_huge_page_size respectively?

File: qemu/tests/numa_memdev_options.py [hunk 15]
```diff
@@ -123,3 +154,4 @@ def run(test, params, env):
 
     error_context.context("Check the memory in procfs!", logging.info)
     check_memory_in_procfs(test, params, vm)
+    vm.verify_kernel_crash()
```
Comment (huangyum) [inline qemu/tests/numa_memdev_options.py]: Would you please add vm.wait_for_login() before verify_kernel_crash() to make sure guest boot up successfully? Others look good to me, thanks.

Comment (huangyum) [reply qemu/tests/numa_memdev_options.py]: Or just use vm.verify_dmesg() instead ?

**Comment by MiriamDeng** [comment]
ACK
 (12/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_bind.no_prealloc: PASS (28.76 s)
 (13/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_interleave.prealloc: PASS (60.62 s)
 (14/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_interleave.no_prealloc: PASS (31.11 s)
 (15/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_preferred.prealloc: PASS (69.16 s)
 (16/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_mlock.policy_preferred.no_prealloc: PASS (31.04 s)
 (17/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_default.prealloc: PASS (130.76 s)
 (18/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_default.no_prealloc: PASS (101.73 s)
 (19/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_bind.prealloc: PASS (121.83 s)
 (20/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_bind.no_prealloc: PASS (130.99 s)
 (21/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_interleave.prealloc: PASS (85.99 s)
 (22/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_interleave.no_prealloc: PASS (91.86 s)
 (23/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_preferred.prealloc: PASS (116.91 s)
 (24/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_hugepage.policy_preferred.no_prealloc: PASS (96.49 s)
 (25/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_default.prealloc: PASS (95.41 s)
 (26/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_default.no_prealloc: PASS (113.98 s)
 (27/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_bind.prealloc: PASS (113.26 s)
 (28/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_bind.no_prealloc: PASS (105.58 s)
 (29/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_interleave.prealloc: PASS (118.89 s)
 (30/32) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.7.ppc64le.io-github-autotest-qemu.numa_memdev_options.numa_ram_hugepage.policy_interleave.no_prealloc: P

**Comment by huangyum** [comment]
Acked-by: Yumei Huang <yuhuang@redhat.com>

**Comment by wangxianxian** [comment]
@lolyu  please review it again, thanks

**Comment by lolyu** [comment]
Thanks all.

**Pull request merged**